lerobot/common/datasets/data_buffer.py

```diff
@@ -0,0 +1,729 @@
+
+        # If we have accessed an existing cached data buffer, just return the object as is.
+        if buffer_already_on_disk:
+            return obj
+
+        # Populate the buffer with the data from the dataset.
+        data_dict = {}
+        for k, feature in hf_dataset.features.items():
+            if isinstance(feature, datasets.features.Image):
+                data_dict[k] = np.stack([np.array(pil_img) for pil_img in hf_dataset[k]])
+            elif isinstance(feature, VideoFrame):
+                if decode_video:
+                    # Decode all videos into images.
+                    episode_indices = np.array(hf_dataset["episode_index"])
+                    timestamps = np.array(hf_dataset["timestamp"])
+                    all_imgs = []
+                    for episode_index in tqdm(
+                        np.unique(episode_indices), desc=f"Decoding videos for {k}", disable=inside_slurm()
+                    ):
+                        episode_data_indices = np.where(episode_indices == episode_index)[0]
+                        episode_timestamps = timestamps[episode_indices == episode_index]
+                        episode_imgs = decode_video_frames_torchvision(
+                            videos_path.parent / hf_dataset[k][episode_data_indices[0]]["path"],
+                            episode_timestamps,
+                            1 / lerobot_dataset_info["fps"] - 1e-4,
+                            to_pytorch_format=False,
+                        )
+                        all_imgs.extend(episode_imgs)
+                    data_dict[k] = np.stack(all_imgs)
+                else:
+                    data_dict[k] = np.stack(
+                        [np.array(dct["path"], dtype=f"S{MAX_VIDEO_PATH_LENGTH}") for dct in hf_dataset[k]]
+                    )
+            elif isinstance(feature, datasets.features.Sequence):
+                data_dict[k] = np.array(hf_dataset[k], dtype=np.dtype(feature.feature.dtype))
+            elif isinstance(feature, datasets.features.Value):
+                data_dict[k] = np.array(hf_dataset[k], dtype=np.dtype(feature.dtype))
+            else:
+                raise NotImplementedError(f"feature type {type(feature)} is not handled.")
+
+        obj.add_episodes(data_dict)  # note this must happen after setting _is_video_dataset.
+        # Symlink vidoes if needed.
+        if obj._is_video_dataset and not obj._videos_dir.exists():
+            os.symlink(videos_path.absolute(), obj._videos_dir)
+        return obj
+
+    @staticmethod
+    def _default_storage_dir_from_huggingface_hub(repo_id: str, fingerprint: str, decode_video: bool) -> Path:
+        """Create the default storage directory used for the `from_huggingface_hub` method.
+
+        Note: This method is really meant for development / testing.
+        """
+        return Path(f"/tmp/{repo_id}_{fingerprint}{'_decoded' if decode_video else ''}")
+
+
+def compute_sampler_weights(
```

Comment:
FYI: This will look very different once we drop `LeRobotDataset`, has decent test coverage for now, and was already here before the PR. Feel free to skim or skip.